File: Interpreter/elements/lexer.py
from elements import Token
from elements import TokenType
from emoji import UNICODE_EMOJI

color_dict = {
    "🟦": "Blue",
    "🟥": "Red",
    "🟩": "Green",
}

bool_dict = {
    "👍": True,
    "👎": False,
}

class Lexer(object):
    def __init__(self, text):
        # client string input, e.g. "4 + 2 * 3 - 6 / 2"
        self.text = text
        # self.pos is an index into self.text
        self.pos = 0
        self.token_type_values = TokenType.get_values()
        self.current_char = self.text[self.pos]
        self.current_line_number = 1

    def error(self):
        raise Exception('{} is an invalid character at line: {}'.format(self.current_char,self.current_line_number))

    def advance(self):
        """Advance the `pos` pointer and set the `current_char` variable."""
        self.pos += 1
        if self.pos > len(self.text) - 1:
            self.current_char = None  # Indicates end of input
        else:
            self.current_char = self.text[self.pos]

    def peek(self):
        peek_pos = self.pos + 1
        if peek_pos > len(self.text) - 1:
            return None
        else:
            return self.text[peek_pos]

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()

    def skip_comment(self):
        while self.current_char != '\n':
            self.advance()

    def number(self):
        """Return a (multidigit) integer or float consumed from the input."""
        result = ''
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()

        if self.current_char == '.':
            result += self.current_char
            self.advance()

            while (
                    self.current_char is not None and
                    self.current_char.isdigit()
            ):
                result += self.current_char
                self.advance()

            token = Token('REAL_CONST', float(result), self.current_line_number)
        else:
            token = Token('INTEGER_CONST', int(result), self.current_line_number)

        return token

    def _id(self):
        """Handle identifiers """
        result = ''
        while self.current_char is not None and (self.current_char.isalnum() or self.current_char in UNICODE_EMOJI):
            result += self.current_char
            self.advance()
        return Token("ID", result, self.current_line_number)

    def get_next_token(self):
        """Lexical analyzer (also known as scanner or tokenizer)
        This method is responsible for breaking a sentence
        apart into tokens. One token at a time.
        """
        while self.current_char is not None:

            if self.current_char == " ":
                self.skip_whitespace()
                continue

            if self.current_char.isdigit():
                return self.number()

            if self.current_char in color_dict:
                token = Token("COLOR", color_dict[self.current_char], self.current_line_number)
                self.advance()
                return token

            if self.current_char in bool_dict:
                token = Token("BOOL_CONST", bool_dict[self.current_char], self.current_line_number)
                self.advance()
                return token

            token = self._read_double_char_symbol()
            if token is not None:
                return token

            token = self._read_single_char_symbol()
            if token is not None:
                if token.type == TokenType.COMMENT:
                    self.skip_comment()
                    continue
                return token

            if self.current_char.isalpha() or self.current_char in UNICODE_EMOJI:
                return self._id()

            self.error()

    def _read_double_char_symbol(self):
        next_char = self.peek()
        if next_char is None:
            return None
        symbol = self.current_char+next_char
        token = None
        if symbol in self.token_type_values:
            token = Token(symbol, symbol, self.current_line_number)
            self.advance()
            self.advance()
        return token

    def _read_single_char_symbol(self):
        token = None
        if self.current_char in self.token_type_values:
            token = Token(self.current_char, self.current_char, self.current_line_number)
            if token.type == TokenType.LINEBREAK:
                self.current_line_number += 1
            self.advance()
        return token


    def lex(self):
        self.lexed_text = []
        self.lexed_text.append(Token('PROGRAM', 'PROGRAM',0))
        while self.current_char is not None:
            token = self.get_next_token()
            if token is not None:
                self.lexed_text.append(token)
        self.lexed_text.append(Token("EOF", "EOF", self.current_line_number+1))


    def __repr__(self):
        return self.lexed_text.__repr__()

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/interpreter/Lexer.kt
package com.dji.emojibleandroid.interpreter

import com.dji.emojibleandroid.utils.EmojiUtils
import java.lang.Exception

class Lexer(val text: String) {
    var pos: Int = 0
    var tokenTypeValues = TokenType.getValues()
    var currentLineNumber = 1
    var currentChar: String? = text[pos].toString()
    var lexedText = mutableListOf<Token>()

    companion object {
        private val colorDict = mapOf("🟦" to "Blue", "🟥" to "Red", "🟩" to "Green")
        private val boolDict = mapOf("👍" to true, "👎" to false)
    }

    private fun error() {
        throw Exception("$currentChar is an invalid character")
    }

    private fun advance() {
        pos += 1
        currentChar = if (pos > text.length - 1) {
            null
        } else {
            text[pos].toString()
        }
    }

    private fun peek(): String? {
        val peekPos = pos + 1
        if (peekPos > text.length - 1) {
            return null
        }
        return text[peekPos].toString()
    }

    private fun skipWhitespace() {
        while (currentChar != null && currentChar!![0].isWhitespace()) {
            advance()
        }
    }

    private fun skipComment() {
        while (currentChar != null && currentChar != "\n") {
            advance()
        }
    }

    private fun number(): Any {
        var result = ""
        while (currentChar != null && currentChar!![0].isDigit()) {
            result += currentChar
            advance()
        }
        return if (currentChar == ".") {
            result += currentChar
            advance()
            while (currentChar != null && currentChar!![0].isDigit()) {
                result += currentChar
                advance()
            }
            Token("REAL_CONST", result.toFloat(), currentLineNumber)
        } else {
            Token("INTEGER_CONST", result.toInt(), currentLineNumber)
        }
    }

    private fun id(): Any {
        var result = ""
        var doubleCharEmoji = currentChar + peek()
        while (currentChar != null &&
            (currentChar!![0].isLetterOrDigit()
                    || EmojiUtils.containsEmoji(currentChar!!)
                    || EmojiUtils.containsEmoji(doubleCharEmoji))
        ) {
            if (EmojiUtils.containsEmoji(doubleCharEmoji)) {
                result += doubleCharEmoji
                advance()
                advance()
                doubleCharEmoji = currentChar + peek()
            } else {
                result += currentChar
                advance()
            }
        }
        return Token(TokenType.ID.value, result, currentLineNumber)
    }

    private fun getNextToken(): Any {
        while (currentChar != null) {

            if (currentChar == " ") {
                skipWhitespace()
                continue
            }

            if (currentChar!![0].isDigit()) {
                return number()
            }

            if (currentChar + peek() in colorDict.keys) {
                val token = Token("COLOR", colorDict[currentChar + peek()], currentLineNumber)
                advance()
                advance()
                return token
            }

            if (currentChar + peek() in boolDict.keys) {
                val token = Token("BOOL_CONST", boolDict[currentChar + peek()], currentLineNumber)
                advance()
                advance()
                return token
            }

            var token = readDoubleCharSymbol()
            if (token != null) {
                if ((token as Token).type == TokenType.COMMENT) {
                    skipComment()
                    continue
                }
                return token
            }

            token = readSingleCharSymbol() as Token?
            if (token != null) {
                if (token.type == TokenType.COMMENT) {
                    skipComment()
                    continue
                }
                return token
            }

            val doubleCharEmoji = currentChar + peek()
            if (currentChar!![0].isLetter() || EmojiUtils.containsEmoji(currentChar!!) || EmojiUtils.containsEmoji(
                    doubleCharEmoji
                )
            ) return id()

            error()
        }
        return Token("EOF", null, currentLineNumber)
    }

    private fun readDoubleCharSymbol(): Any? {
        val nextChar = peek() ?: return null
        val symbol = currentChar + nextChar
        var token: Token? = null
        if (symbol in tokenTypeValues) {
            token = Token(symbol, symbol, currentLineNumber)
            advance()
            advance()
        }
        return token
    }

    private fun readSingleCharSymbol(): Any? {
        var token: Token? = null
        if (currentChar in tokenTypeValues) {
            token = currentChar?.let { Token(it, currentChar, currentLineNumber) }
            if (token?.type == TokenType.LINEBREAK) {
                currentLineNumber += 1
            }
            advance()
        }
        return token
    }

    fun lex() {
        lexedText.add(Token("PROGRAM", "PROGRAM", 0))
        while (currentChar != null) {
            lexedText.add(getNextToken() as Token)
        }
        if (lexedText[lexedText.size - 1].type != TokenType.EOF) {
            lexedText.add(Token("EOF", null, currentLineNumber + 1))
        }
    }
}